packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -420,6 +427,64 @@ library ExchangeBlocks
         emit BlockCommitted(S.blocks.length - 1, publicDataHash);
     }
 
+    function validateConditionalTransfers(
+        ExchangeData.State storage S,
+        uint32 newBlockIdx,
+        bytes  memory data,
+        bytes  memory auxiliaryData
+        )
+        private
+    {
+        // Read how many conditional transfers are included in the block
+        uint numConditionalTransfers;
+        assembly {
+            numConditionalTransfers := and(mload(add(data, 104)), 0xFFFFFFFF)
+        }
+        require(auxiliaryData.length == numConditionalTransfers * 6, "INVALID_AUXILIARYDATA");
+
+        // Run over all conditional transfers
+        uint numBytesPerTransfer = 13;
+        for (uint i = 0; i < auxiliaryData.length; i += 6) {
+            // auxiliaryData contains for each conditional transfer:
+            // - Transfer index: 2 bytes (the position of the conditional transfer inside data)
+            // - Salt: 4 bytes (the salt of the transfer)
+            uint index;
+            uint salt;
+            assembly {
+                index := and(mload(add(auxiliaryData, add(i, 2))), 0xFFFF)
+                salt := and(mload(add(auxiliaryData, add(i, 6))), 0xFFFFFFFF)
+            }
+            uint offset = 104 + index * numBytesPerTransfer;
+            uint transferData;
+            assembly {
+                transferData := and(mload(add(data, offset)), 0xFFFFFFFFFFFFFFFFFFFFFFFFFF)
+            }
+
+            // Check that this is a conditional transfer
+            require((transferData & 0xFF) == 1, "INVALID_AUXILIARYDATA");
+
+            // The key of the transfer is the combination of the transfer DA data and the salt
+            uint key = (transferData << 32) | salt;
+            ExchangeData.ConditionalTransferState memory transferState = S.conditionalTransfers[key];
+            require(transferState.approved, "TRANSFER_UNAUTHORIZED");
+
+            // Make sure the request was either not used yet, or used in a block that was reverted
+            if (transferState.blockIdx != 0 && transferState.blockIdx < S.blocks.length) {
+                // A block exists, but the block could have been changed.
+                // Check if the block has a newer Ethereum block number
+                require(
+                    S.blocks[transferState.blockIdx].ethereumBlockNumber > transferState.ethereumBlockNumber,
```

Comment:
Now I understand the purpose of keeping the record of Ethereum block numbers in a virtual block.
Just to straight my mind:
1.  if `transferState.blockIdx == 0` => fine, this state is not used, so we can use it.
1.  otherwise, the block has been used, but if `transferState.blockIdx >= S.blocks.length`, it means the state is used by a block that has been reverted;
1.  otherwise, the state is used by a previous block, but that block on the current virtual chain may be different from the block that actually consumed the state in the first place.... to check if this is the case, we use the Ethereum block number, if for the same virtual block index, the Ethereum block height is larger than the one recorded in the state, we say the vblock is reverted so the state can be reused.

We are actually assuming that **a reverted vblock must have a bigger Ethereum height than the one it replaced**.

Reply:
Another approach is to record the Ethereum block hash in our virtual block, then the above assumption doesn't have to hold. It will consume more gas though. 

Is it technically possible for the operator to commit a vblock and then immediately revert it in the same Ethereum block?  This will invalidate the above assumption. Therefore we may need to do certain checking in the `revertBlock` funciton.

Reply:
> Now I understand the purpose of keeping the record of Ethereum block numbers in a virtual block.
> Just to straight my mind:
> 
>     1. if `transferState.blockIdx == 0` => fine, this state is not used, so we can use it.
> 
>     2. otherwise, the block has been used, but if `transferState.blockIdx >= S.blocks.length`, it means the state is used by a block that has been reverted;
> 
>     3. otherwise, the state is used by a previous block, but that block on the current virtual chain may be different from the block that actually consumed the state in the first place.... to check if this is the case, we use the Ethereum block number, if for the same virtual block index, the Ethereum block height is larger than the one recorded in the state, we say the vblock is reverted so the state can be reused.

Totally correct.

> We are actually assuming that **a reverted vblock must have a bigger Ethereum height than the one it replaced**.

True.





> Another approach is to record the Ethereum block hash in our virtual block, then the above assumption doesn't have to hold. It will consume more gas though.

Yes, I think that would also work.


> Is it technically possible for the operator to commit a vblock and then immediately revert it in the same Ethereum block? This will invalidate the above assumption. Therefore we may need to do certain checking in the `revertBlock` funciton.

Yes, it is currently possible to revert a block immediately (the delay until the proof generation period has expired was removed). The check is `require(S.blocks[transferState.blockIdx].ethereumBlockNumber > transferState.ethereumBlockNumber)` so the operator wouldn't even be able to reuse the request in a new block in the same Ethereum block, in this special case the operator needs to wait at least 1 Ethereum block before the conditional transfer can be reused in a new block, so I don't think it's a problem. But I didn't think of this and an extra check in `revertBlock` is a good idea.